File: oif_impl/python/_serialization.py
from io import BytesIO

import msgpack


def deserialize(sd: bytes) -> dict:
    """Transform serialized dictionary `sd` to native Python dict."""

    bytes_stream = BytesIO(sd)
    unpacker = msgpack.Unpacker(bytes_stream, raw=False)

    data = []

    for datum in unpacker:
        data.append(datum)

    assert len(data) % 2 == 0, "Malformed serialized config dict"

    resultant_dict = {}
    i = 0
    while i < len(data):
        key = data[i]
        value = data[i + 1]

        resultant_dict[key] = value
        i += 2

    return resultant_dict

File: oif_impl/lang_julia/serialization.jl
module OIFSerialization
export deserialize
import MsgPack

function deserialize(sd::Ptr{Cchar})::Dict
    sd_str = unsafe_string(sd)
    io = IOBuffer(sd_str)

    data = MsgPack.unpack(io)
    @assert length(data) % 2 == 0

    resultant_dict = Dict()
    i = 1
    while i <= length(data)
        key = data[i]
        value = data[i + 1]

        resultant_dict[key] = value
        i += 2
    end

    return resultant_dict
end
end
